Pool without keyspace but with close event

ctavan opened the pull request:
When adding a test case for the keyspace-less connects I stumbled upon the problem, that a pool doesn't emit a close event as soon as all client connections have been closed.

This patch adds both, keyspaceless connects, and emission of closeevents by pools.

However something still goes wrong in the thrift test where a connection is now opened, then closed, then opened again and closed again. The test times out in the and. Apparently some connection still remains open.

Could anyone look into this? I'm stuck.

ctavan committed:
Allow connecting without defining a keyspace

- Make Pool emit close event
- Also add test that tests connecting to cassandra without specifying a
  Keyspace
- Don't create monitoring interval if the pool is already being closed
- Stick to implicit coding conventions

devdazed commented:
i just had a look. it seems the connection is remaining open, event though it is emitting close.  this may be a problem in the underlying thrift driver.

devdazed commented:
actually, it seems like the underlying `net.Connection` is closing.  I'm not sure what is causing the timeout

below is the status after the test has finished.  `_connection.connection` is the `net.Connection` object.  as you can see, the handle is `null` and it isn't readable or writable.  Lets be sure that all intervals have been cleared

```
{ port: '9160',
  host: 'localhost',
  timeout: 3000,
  user: 'test',
  password: 'test1233',
  keyspace: 'system',
  ready: false,
  _events: { error: [Function], close: [Function: closed] },
  _connection: 
   { connection: 
      { _handle: null,
        _pendingWriteReqs: 0,
        _flags: 3,
        _connectQueueSize: 0,
        destroyed: true,
        errorEmitted: false,
        bytesRead: 12432,
        bytesWritten: 6936,
        allowHalfOpen: undefined,
        _connecting: false,
        writable: false,
        _events: [Object],
        readable: false,
        _idleNext: null,
        _idlePrev: null,
        _idleTimeout: -1,
        frameLeft: 0,
        framePos: 0,
        frame: null,
        _connectQueue: null },
     options: {},
     transport: { [Function] receiver: [Function] },
     protocol: [Function],
     offline_queue: [],
     connected: true,
     host: 'localhost',
     port: '9160',
     _events: { error: [Function], close: [Function], connect: [Function] },
     client: { output: [Object], pClass: [Function], seqid: 51, _reqs: {} } },
  _client: 
   { output: 
      { inBuf: <Buffer >,
        outBuffers: [],
        outCount: 0,
        readPos: 0,
        onFlush: [Function] },
     pClass: [Function],
     seqid: 51,
     _reqs: {} } }
```

ctavan commented:
The problem was, that `Pool.monitorConnections()` was called after the `close` event has been emitted when a pool is connected and then immediately closed. This resulted in a superfluous `retryInterval` being created. I've fixed the problem by adding a new property `closing` to the pool that gets set as soon as `Pool.close()` is called and that prevents the pool from trying to monitor connections.

I've squashed all commits into a single commit.

ctavan commented:
I've just rebased this branch.

ctavan commented:
Rebased against current upstream master again.

ctavan commented:
What do you think about this one now?

devdazed commented:
looks good
